3-Solving-Problems-By-Searching/poisson_distribution.js

```diff
@@ -0,0 +1,78 @@
+function poissonDistribution(height, width, radius) {
+  let sample = poissonDiscSampler(width, height, radius);
+  let nodes = [];
+  let s = sample();
+  let grid = [];
+  while (s) {
+    nodes.push(s.point);
+    grid = s.grid;
+    s = sample();
+  }
+  return [nodes, grid];
+}
+
+function poissonDiscSampler(width, height, radius) {
```

Comment:
**No**, sorry, adding the license would not be enough. The short version is: in order to use other people's code, even open source, you have to follow all the rules, and GPL has different rules than the MIT license. Adding the license isn't enough to follow the GPL rules. There are several issues here.

1. Any code you check into the repository is by default copyrighted *by you*. (Berne Convention and other laws). If you checked this in, you would be claiming that you are the author. But **you are not the author**. This is true *even of open source*. We would need to put this into a `third-party/` folder along with the copyright and license information, so that we are making it clear that it is not code you wrote, and you are not claiming copyright over it.
2. Open source code is copyrighted just like any other code. You cannot use copyrighted code by default. The open source license grants permission to use it *under certain requirements*. You can only use it in if you follow those requirements.
3. The license lists the conditions under which you may use it. The AIMA code is under the [MIT license](https://en.wikipedia.org/wiki/MIT_License). The poisson disc code is under the [GPL 3 license](https://www.gnu.org/licenses/gpl.html). Both of these licenses require that they retain the original copyright and license, but poisson_distribution.js does not do either. That means it's in violation of the license and it cannot be used. Adding the license and putting it into a third-party/ folder would meet this requirement.
4. GPL also requires that all *other* source code in the program also be released under the GPL license. That means if you check in poisson_distribution.js, *the rest of the project* has to be GPL. That means we would need to go to each of the contributors to aima-javascript and ask them to re-release the code under the GPL license. So adding the license and putting it into third-party/ is not enough to use GPL code, unless we re-license everything else as GPL, and I don't think we're going to do that.

That said, it is a *good idea* to use libraries. We just have to use them in ways they allow. You could either write a separate program that uses the GPL library and writes out JSON, and then you would mark that program as GPL also (without making the rest of the project GPL), or you could use a library with a different license, such as [poisson-disk-sampling](https://github.com/kchapelier/poisson-disk-sampling), which is under the MIT license.

I think the cleanest thing to do would be to write a separate program that runs in node.js that uses the poisson-disk-sampling library, and have it write out JSON.

A simpler option would be to make a grid instead of the poisson disk graph. Offset grids (every other row is shifted by 1/2 column) and jittered grids (every point is moved by a small random amount) would provide enough variety to make the graph look nice.